internal/rukpak/preflights/crdupgradesafety/crdupgradesafety_test.go

```diff
@@ -0,0 +1,203 @@
+
+	"github.com/operator-framework/operator-controller/internal/rukpak/preflights/crdupgradesafety"
+	"github.com/operator-framework/operator-controller/internal/rukpak/util"
+)
+
+type MockCRDGetter struct {
+	oldCrd *apiextensionsv1.CustomResourceDefinition
+	getErr error
+	apiextensionsv1client.CustomResourceDefinitionInterface
+}
+
+func (c *MockCRDGetter) Get(ctx context.Context, name string, options metav1.GetOptions) (*apiextensionsv1.CustomResourceDefinition, error) {
+	return c.oldCrd, c.getErr
+}
+
+func newMockPreflight(crd *apiextensionsv1.CustomResourceDefinition, err error, customValidator *kappcus.Validator) *crdupgradesafety.Preflight {
+	var preflightOpts []crdupgradesafety.Option
+	if customValidator != nil {
+		preflightOpts = append(preflightOpts, crdupgradesafety.WithValidator(customValidator))
+	}
+	return crdupgradesafety.NewPreflight(&MockCRDGetter{
+		oldCrd: crd,
+		getErr: err,
+	}, preflightOpts...)
+}
+
+func getCrdFromManifestFile(t *testing.T, oldCrdFilePath string) *apiextensionsv1.CustomResourceDefinition {
+	if oldCrdFilePath == "" {
+		return nil
+	}
+	relObjects, err := util.ManifestObjects(strings.NewReader(getManifestString(t, oldCrdFilePath)), "old")
+	require.NoError(t, err)
+
+	newCrd := &apiextensionsv1.CustomResourceDefinition{}
+	for _, obj := range relObjects {
+		if obj.GetObjectKind().GroupVersionKind() != apiextensionsv1.SchemeGroupVersion.WithKind("CustomResourceDefinition") {
+			continue
+		}
+		uMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+		require.NoError(t, err)
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(uMap, newCrd)
+		require.NoError(t, err)
+	}
+	return newCrd
+}
+
+func getManifestString(t *testing.T, filePath string) string {
+	buff, err := os.ReadFile(filePath)
+	require.NoError(t, err)
+	return string(buff)
+}
+
+// TestInstall exists only for completeness as Install() is currently a no-op. It can be used as
+// a template for real tests in the future if the func is implemented.
+func TestInstall(t *testing.T) {
+	preflight := newMockPreflight(nil, nil, nil)
+	require.Nil(t, preflight.Install(context.Background(), nil))
+}
+
+func TestUpgrade(t *testing.T) {
+	tests := []struct {
+		name          string
+		oldCrdPath    string
+		validator     *kappcus.Validator
+		release       *release.Release
+		wantErrMsgs   []string
+		wantCrdGetErr error
+	}{
+		{
+			name: "nil release",
+		},
+		{
+			name: "release with no objects",
+			release: &release.Release{
+				Name: "test-release",
+			},
+		},
+		{
+			name: "release with invalid manifest",
+			release: &release.Release{
+				Name:     "test-release",
+				Manifest: "abcd",
+			},
+			wantErrMsgs: []string{"json: cannot unmarshal string into Go value of type unstructured.detector"},
+		},
+		{
+			name: "release with no CRD objects",
+			release: &release.Release{
+				Name:     "test-release",
+				Manifest: getManifestString(t, "../../../../testdata/manifests/no-crds.json"),
```

Comment:
Nit: I'm personally not a huge fan of unit tests having dependencies on a file existing. I find it harder to grok what the unit test is trying to achieve when I have to jump back and forth between the unit test code and a file with $content

Instead of having JSON files (which are harder to read than YAML, IMO), what do you think about having the Go types for the manifests defined in the test itself and then at test runtime building the release with that data?

Reply:
I think we may differ in opinion here :sweat_smile: I personally don't like having to scroll through hundreds of lines of code for static structs and strings in test files; putting them in separate json files makes the test code cleaner IMHO. 

This doesn't help with your nit at all but I can at least make a `const` for the folder location so we don't see the `.../../../../` all over the place...

Reply:
@dtfranz No need to do anything for the nit unless you'd like. Just something I felt worth calling out in case others felt more strongly one way or the other.

Reply:
I'd prefer YAML files as well, even though the JSONs are kinda-sorta formatted like YAML...